guidance_programs/fewshot_cot_as_conversation_ensemble.py

```diff
@@ -0,0 +1,157 @@
+        if i_swap < len(sorted_swaps) and i == sorted_swaps[i_swap]:
+            result.append(line[sorted_swaps[i_swap] + 1])
+        elif i_swap < len(sorted_swaps) and i == sorted_swaps[i_swap] + 1:
+            result.append(line[sorted_swaps[i_swap]])
+            i_swap += 1
+        else:
+            result.append(line[i])
+    return result
+
+
+def plain_hunt_generator(starting_line: list[T]) -> Iterator[T]:
+    first_element = starting_line[0]
+    swaps_A = list(range(0, len(starting_line) - (len(starting_line) % 2), 2))
+    swaps_B = list(range(1, len(starting_line) - 1, 2))
+    all_swaps = [swaps_A, swaps_B]
+    current = [x for x in starting_line]
+    line_count = 0
+    yield current
+    while True:
+        current = apply_swaps(current, all_swaps[line_count % len(all_swaps)])
+        yield current
+        line_count += 1
+        if current[0] == first_element:
+            break
+
+
+NUM_PERMUTATIONS = 5
+
+
+@guidance
+def few_shot_cot_multiple_choice(
+    lm: guidance.models.Chat,
+    question: str,
+    choices: list[str],
+    fewshot_examples: list[dict[str, any]],
+    permutation: list[int],
+):
+    # Some general instruction to the model
+    with system():
+        lm += textwrap.dedent(
+            """Answer the following multiple choice **Question**.
+            First, think step by step and write an **Explanation** for reasoning through the question.
+            Then, when prompted by the user for a **Final Answer**, analyze your explanation and write just the number of the correct answer.
+            Do not say the final answer until the user asks for it."""
+        )
+
+    for example in fewshot_examples:
+        with user():
+            lm += "**Question**\n"
+            lm += example["question"] + "\n"
+            for i, choice in enumerate(example["choices"]):
+                lm += f"{i} : {choice}\n"
+            lm += "**Explanation**"
+
+        with assistant():
+            lm += example["chain_of_thought"]
+
+        with user():
+            lm += f"**Final Answer**"
+
+        with assistant():
+            lm += str(example["correct_answer"])
+
+    with user():
+        lm += question + "\n"
+        for i in range(len(choices)):
+            lm += f"{i}: {choices[permutation[i]]}"
+        lm += "**Explanation**"
+
+    with assistant():
+        lm += gen(name=COT_KEY)
+
+    with user():
+        lm += f"**Final Answer**"
+
+    with assistant():
+        lm += select([str(i) for i in range(len(choices))], name=ANSWER_KEY)
+
+    return lm
+
+
+def guidance_generation(
+    lm: guidance.models.Chat,
+    input: Dict[str, Any],
+    common: list[dict[str, Any]] | None = None,
+) -> Dict[str, Any]:
+    _logger.debug("Starting guidance_generation")
+    assert common is None, "Unexpected common data"
+
+    votes = [0, 0, 0, 0]
+    cots = []
+    generator = plain_hunt_generator(list(range(len(input["choices"]))))
+    for i in range(NUM_PERMUTATIONS):
```

Comment:
As we discussed yesterday @Harsha-Nori , the `guidance` program is just Python, so we can sneak things like this in :-)